File: client/extension/graphTypes.ts
/**
 * Types for the graph data returned by getGraphData command
 */

/**
 * Represents a location in a file
 */
export interface GraphLocation {
    /** File path with forward slashes */
    filename: string;
    /** Line number (1-based) */
    line: number;
    /** Column number (1-based) */
    column: number;
}

/**
 * Represents a reference between graph nodes
 */
export interface GraphReference {
    /** The key/id of the referenced node */
    key: string;
    /** Whether this is an outgoing reference */
    isOutgoing: boolean;
    /** Optional label for the reference */
    label?: string;
}

/**
 * Represents a node in the graph
 */
export interface GraphNode {
    /** Unique identifier for the node */
    id: string;
    /** Display name for the node */
    name?: string;
    /** List of references to other nodes */
    references: GraphReference[];
    /** Location of the node in a file */
    location?: GraphLocation;
    /** Documentation for the node */
    documentation?: string;
    /** Additional details as key-value pairs */
    details?: GraphNodeDetail[];
    /** Whether this is a primary node */
    isPrimary: boolean;
    /** Type of the entity */
    entityType: string;
    /** Display name for the entity type */
    entityTypeDisplayName?: string;
    /** Abbreviation for the node */
    abbreviation?: string;
}

/**
 * Represents a detail entry for a graph node
 */
export interface GraphNodeDetail {
    /** Key for the detail */
    key: string;
    /** Values associated with the key */
    values: string[];
}

/**
 * Complete graph data returned by getGraphData
 */
export type GraphData = GraphNode[];

/**
 * Wrapper function for getGraphData command
 * @param entityType The type of entity to get graph data for
 * @param depth The depth of connections to include
 * @returns Promise with the graph data
 */
export async function getGraphData(entityType: string, depth: number): Promise<GraphData> {
    const commands = await import('vscode').then(vscode => vscode.commands);
    const result = await commands.executeCommand<any[]>("getGraphData", entityType, depth);
    return result as GraphData;
}

File: src/Main/GraphTypes.fs
module Main.GraphTypes

open CWTools.Utilities.Position

/// Represents a reference between graph nodes
type GraphReference =
    {
        /// The key/id of the referenced node
        key: string
        /// Whether this is an outgoing reference
        isOutgoing: bool
        /// Optional label for the reference
        label: string option
    }

/// Represents a node in the graph
type GraphNode =
    {
        /// Unique identifier for the node
        id: string
        /// Display name for the node
        displayName: string option
        /// List of references to other nodes
        references: GraphReference list
        /// Location of the node in a file
        location: range option
        /// Documentation for the node
        documentation: string option
        /// Additional details as key-value pairs
        details: Map<string, string list> option
        /// Whether this is a primary node
        isPrimary: bool
        /// Type of the entity
        entityType: string
        /// Display name for the entity type
        entityTypeDisplayName: string option
        /// Abbreviation for the node
        abbreviation: string option
    }

/// Complete graph data returned by getGraphData
type GraphData = GraphNode list

/// Converts a GraphNode to a JSON representation
let graphNodeToJson (node: GraphNode) =
    let convRangeToJson (loc: range) =
        [| "filename", FSharp.Data.JsonValue.String(loc.FileName.Replace("\\", "/"))
           "line", FSharp.Data.JsonValue.Number(loc.StartLine |> decimal)
           "column", FSharp.Data.JsonValue.Number(loc.StartColumn |> decimal) |]
        |> FSharp.Data.JsonValue.Record

    let referenceToJson (ref: GraphReference) =
        [| Some("key", FSharp.Data.JsonValue.String ref.key)
           Some("isOutgoing", FSharp.Data.JsonValue.Boolean ref.isOutgoing)
           ref.label |> Option.map (fun l -> "label", FSharp.Data.JsonValue.String l) |]
        |> Array.choose id
        |> FSharp.Data.JsonValue.Record

    let detailsToJson (details: Map<string, string list>) =
        details
        |> Map.toArray
        |> Array.map (fun (k, vs) ->
            FSharp.Data.JsonValue.Record
                [| "key", FSharp.Data.JsonValue.String k
                   "values",
                   (vs
                    |> Array.ofList
                    |> Array.map FSharp.Data.JsonValue.String
                    |> FSharp.Data.JsonValue.Array) |])
        |> FSharp.Data.JsonValue.Array

    [| Some("id", FSharp.Data.JsonValue.String node.id)
       node.displayName
       |> Option.map (fun s -> ("name", FSharp.Data.JsonValue.String s))
       Some("references", FSharp.Data.JsonValue.Array(node.references |> Array.ofList |> Array.map referenceToJson))
       node.location |> Option.map (fun loc -> "location", convRangeToJson loc)
       node.documentation
       |> Option.map (fun s -> "documentation", FSharp.Data.JsonValue.String s)
       node.details |> Option.map (fun m -> "details", detailsToJson m)
       Some("isPrimary", FSharp.Data.JsonValue.Boolean node.isPrimary)
       Some("entityType", FSharp.Data.JsonValue.String node.entityType)
       node.entityTypeDisplayName
       |> Option.map (fun s -> ("entityTypeDisplayName", FSharp.Data.JsonValue.String s))
       node.abbreviation
       |> Option.map (fun s -> ("abbreviation", FSharp.Data.JsonValue.String s)) |]
    |> Array.choose id
    |> FSharp.Data.JsonValue.Record

/// Converts GraphData to a JSON representation
let graphDataToJson (data: GraphData) =
    data |> List.map graphNodeToJson |> Array.ofList |> FSharp.Data.JsonValue.Array
